include/accessor.h

```diff
@@ -254,12 +254,17 @@ class accessor<DataT, Dims, Mode, target::device> : public detail::accessor_base
 	}
 
 	template <access_mode M = Mode>
-	inline std::enable_if_t<detail::access::mode_traits::is_producer(M), DataT&> operator[](const id<Dims>& index) const {
-		return m_device_ptr[get_linear_offset(index)];
-	}
-
-	template <access_mode M = Mode>
-	inline std::enable_if_t<detail::access::mode_traits::is_pure_consumer(M), const DataT&> operator[](const id<Dims>& index) const {
+	inline std::conditional_t<detail::access::mode_traits::is_producer(M), DataT&, const DataT&> operator[](const id<Dims>& index) const {
+#if CELERITY_ACCESSOR_BOUNDARY_CHECK
+		if(((index - m_backing_buffer_offset) >= m_backing_buffer_range) != detail::id_cast<Dims>(id<3>(false, false, false))) {
```

Comment:
Should we be checking against the backing buffer range here, or against what the user has declared they are going to access? I'm thinking the latter! (This would require adding one additional field to the accessor).